test/integration/extnet_test.go

```diff
@@ -0,0 +1,258 @@
+//go:build integration
+
+/*
+Copyright 2020 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"encoding/json"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/minikube/cmd/minikube/cmd"
+)
+
+// TestExtNet tests minikube external network functionality
+func TestExtNet(t *testing.T) {
+	driverArg, driverArgPresent := DriverArg()
+	t.Logf("running with runtime:%s DriverArg:%s goos:%s goarch:%s", ContainerRuntime(), driverArg, runtime.GOOS, runtime.GOARCH)
+	if driverArgPresent && driverArg != "docker" {
+		t.Skip("skipping: only docker driver supported")
+	}
+
+	type validateFunc func(context.Context, *testing.T, string, string)
+	profile := UniqueProfileName("extnet")
+	ctx, cancel := context.WithTimeout(context.Background(), Minutes(10))
+	defer Cleanup(t, profile, cancel)
+
+	containerRuntime := ContainerRuntime()
+	switch containerRuntime {
+	case "containerd":
+	    t.Skip("skipping: access direct test is broken on windows: https://github.com/kubernetes/minikube/issues/8304")
+	case "crio":
+	    t.Skip("skipping: access direct test is broken on windows: https://github.com/kubernetes/minikube/issues/8304")
+	case "docker":
+	}
+
+	extnetNetworkName := fmt.Sprintf("%s-%s", "network-extnet", fmt.Sprintf("%06d", time.Now().UnixNano()%1000000))
+
+	// Serial tests
+	t.Run("serial", func(t *testing.T) {
+		tests := []struct {
+			name      string
+			validator validateFunc
+		}{
+			{"CreateExtnet", createExtnet},
+			{"FreshStart", extnetValidateFreshStart},
+			{"ConnectExtnet", connectExtnet},
+			{"Stop", extnetValidateStop},
+			{"VerifyStatus", extnetValidateStatus},
+			{"Start", extnetValidateStart},
+			{"Delete", extnetValidateDelete},
+		//	{"Fail", fail},
+			{"VerifyDeletedResources", extnetValidateVerifyDeleted},
```

Comment:
I copied this cleaning principle from pause_test.go. I think it's quite nice to test the deletion with a very controlled way. And the Cleanup is called effectively in case of early error. I made some change to be closer to this principle (will talk about this in the main message)